### ISSUE TITLE
Issue with alsa stream

### ISSUE THREAD

**Issue by rad-llc** [badaix/snapcast#827]
I am using this in my snapmaster.conf file running snapserver on raspberry pi 3.

stream = alsa://?name=default&device=hw:0,0&sampleformat=44100:16:2

I am getting this issue (below).  Adding snapserver to the user group audio (as suggested in another issue) is not resolving this:

2021-03-09 16-00-33.474 [Info] (Snapserver) Version 0.24.0, revision 7eff45fd
2021-03-09 16-00-33.474 [Info] (Snapserver) Adding stream: pipe:///tmp/snapfifo?name=Snapserver&sampleformat=44100:16:2
2021-03-09 16-00-33.474 [Info] (Snapserver) Adding stream: alsa://?name=default&device=hw:0,0&sampleformat=44100:16:2
2021-03-09 16-00-33.475 [Notice] (init) Settings file: "/home/pi/.config/snapserver/server.json"
2021-03-09 16-00-33.488 [Info] (Avahi) Adding service 'Snapcast'
2021-03-09 16-00-33.501 [Info] (PcmStream) PcmStream: Snapserver, sampleFormat: 44100:16:2
2021-03-09 16-00-33.501 [Info] (PcmStream) Stream: Snapserver, metadata={
2021-03-09 16-00-33.501 [Info] (PcmStream)     "STREAM": "Snapserver"
2021-03-09 16-00-33.501 [Info] (PcmStream) }
2021-03-09 16-00-33.501 [Info] (Server) onMetaChanged (Snapserver)
2021-03-09 16-00-33.502 [Info] (PipeStream) PipeStream mode: create
2021-03-09 16-00-33.502 [Info] (Server) Stream: {"fragment":"","host":"","path":"/tmp/snapfifo","query":{"chunk_ms":"20","codec":"flac","name":"Snapserver","sampleformat":"44100:16:2"},"raw":"pipe:////tmp/snapfifo?chunk_ms=20&codec=flac&name=Snapserver&sampleformat=44100:16:2","scheme":"pipe"}
2021-03-09 16-00-33.503 [Info] (PcmStream) PcmStream: default, sampleFormat: 44100:16:2
2021-03-09 16-00-33.503 [Info] (PcmStream) Stream: default, metadata={
2021-03-09 16-00-33.503 [Info] (PcmStream)     "STREAM": "default"
2021-03-09 16-00-33.503 [Info] (PcmStream) }
2021-03-09 16-00-33.503 [Info] (Server) onMetaChanged (default)
2021-03-09 16-00-33.503 [Info] (Server) Stream: {"fragment":"","host":"","path":"","query":{"chunk_ms":"20","codec":"flac","device":"hw:0,0","name":"default","sampleformat":"44100:16:2"},"raw":"alsa:///?chunk_ms=20&codec=flac&device=hw:0,0&name=default&sampleformat=44100:16:2","scheme":"alsa"}
2021-03-09 16-00-33.513 [Notice] (Server) Server::start: Can't open device 'hw:0,0', error: No such file or directory
2021-03-09 16-00-33.514 [Error] (Snapserver) Exception: Can't open device 'hw:0,0', error: No such file or directory
2021-03-09 16-00-33.516 [Notice] (Snapserver) Snapserver terminated.

**Comment by kingosticks** [comment]
You've also posted at https://github.com/badaix/snapcast/issues/710#issuecomment-794510002 Which one should someone try to answer? Posting the same thing twice just creates more work so please don't do that.

You need to provide all the requested information from the bug report template. How was this installed? Is this Raspberry Pi OS Desktop?

**Comment by kingosticks** [comment]
Desktop or Lite?

**Comment by radeckellc** [comment]
Please bear with me, very new to linux. The image is raspberry pi lite.  hw:0,0 are the headphones and I am able play sounds to 3.5mm jack referencing hw:0,0 as the alsa device, as well as referencing the same device as default.  As such, I can Bluetooth into the raspberry pi with my phone, and play music through the 3.5 mm jack to a speaker.  I want to send the music stream to all the clients, hence setting up alsa in my snapcaster.conf file, but I get the above error.

**Comment by kingosticks** [comment]
> stream = alsa://?name=default&device=hw:0,0&sampleformat=44100:16:2

Do you mean:
```
source = alsa://?name=default&device=hw:0,0&sampleformat=44100:16:2
```
?

`source` definitions in the `[stream]` section of a snapserver config are for specifying sources of audio data that snapserver will read from. You cannot use the Raspberry Pi's 3.5mm jack ALSA *output* as an ALSA *input*. The error message is correctly telling you there is no input device at hw:0,0. If you want to use a snapserver `alsa` source then take a look at the [documentation](https://github.com/badaix/snapcast/blob/master/doc/configuration.md#alsa) which walks you through creating and using an ALSA loopback device.

**Comment by radeckellc** [comment]
Ok thanks, let me try that.  So I think what you are saying is that HW:0,0 exists but as output and I need to loop it back to an input.  I will look at the documentation and give it a try.  Again, this is new to me.  Let me give that a try and I will report back here.

**Comment by radeckellc** [comment]
Well thank you very much.  I have it working now!  That was it.  I definitely missed that step because it did not mean anything to me.  It is working great.  Now, I need to figure out how to get two bluetooth devices setup (on two separate Ras Pi's to stream music through snapcast without out have two snapservers.  Reason I want this is I need to locate the bluetooth devices in different levels of the home.  Thanks again.

**Comment by radeckellc** [comment]
This is giving me fits.  For some reason it is not working.  I did get the alsa loopback device created and placed it in my .conf file.  No issues anymore when I start the server.  However when I play music now through bluetooth on the server, it is not picking up the loopback device some reason.  On the server (the one with snapserver), It plays on hw:0,0, which are headphones but the loopback does not appear to be working and picked up by the clients.

**Comment by kingosticks** [comment]
> but the loopback does not appear to be working and picked up by the clients.

I don't know how Bluetooth streaming to a Raspberry Pi is configured these days (it seems to change every release) but it probably needs to be explicitly told to use the loopback device for output. If you can't explicitly tell it what device to use then you can try making your new loopback device the default ALSA output device.

**Comment by radeckellc** [comment]
Ugh - I am not sure even how to do this.  I will have to do some digging and report back.

**Comment by radeckellc** [comment]
I am uncertain how change the bluetooth default but it does appear to be defaulting correctly.  I am connected to bluetooth and it no longer plays through to the headphones so it seems to playing through one of the loopback devices.  I do see that the snapsever onstate changes when I toggle play on and of as I am connected to bluetooth.  It is just that I am getting no sound out of the client.

**Comment by rad-llc** [comment]
I never did get this figured out. I am using this script to setup bluetooth - https://github.com/nicokaiser/rpi-audio-receiver (just use the one bluetooth script).

On my snapserver, when I setup alsa looback through sudo modprobe snd-aloop (and also place snd-loop in etc/modules in order for it to survive on boot), my hardware devices become hw:0,0 and hw:0,1 are the loopback devices and hw:1,0 becomes the 3.5mm headphone jack.    

Bluetooth continues to play to hw:0,0 (which was formerly the 3.5mm headphone jack before executing snd-aloop).  I then setup my snapserver.conf  to: 
source = alsa://?name=default&device=hw:0,1&sampleformat=44100:16:2  in order to receive the bluetooth playback and run it through snapserver.

As far as I can tell, snapserver and clients work correctly as I can see activity when I toggle playback off and on through when I am bluetooted into sanpserver.  The issue is that I never get any sound.  It is not a bluetooth issue(as it worked previously when playing through the 3.5mm headphone jack when it was hw:0,0) as I ran tests with aplay to hw:0,0 and arecord of hw:0,1 and substantiated that I can record but have no sound on the recording either.  Another matter is that because bluetooth is is now playing to hw:0,0 (which is no longer the 3.5mm headphone jack), I do not even know if it possible to also direct it to hw:1,0 so I can continue to get sound locally out of the headphone jack.

I have spent hours on this.   The main reason why I want this setup is that I need a bluetooth receiver for other music sources other than Spotify (as I use Raspotify) to play on my various snapclient receivers.  My skills are not deep enough to troubleshoot this.
